rr_iarrc/src/cone_detection/cone_height_detector.cpp

```diff
@@ -0,0 +1,379 @@
+typedef cv::Point Point;
+int low_H, high_H, low_S, low_V;
+int canny_cut_min_threshold, minimum_area_cut;
+double percent_max_distance_transform;
+
+
+// PointCloud Processing
+pcl::PointCloud<pcl::PointXYZ> draw_cone_circles();
+void drawCircle(pcl::PointCloud<pcl::PointXYZ>&, pcl::PointXYZ, double, int);
+void fovCallback(const sensor_msgs::CameraInfoConstPtr& msg);
+void loadCameraFOV(ros::NodeHandle& nh);
+void publish_closest_point();
+
+cv_bridge::CvImagePtr cv_ptr;
+typedef pcl::PointCloud<pcl::PointXYZ> PointCloud;
+double fy, real_cone_height, real_cone_radius;
+double camera_fov_horizontal;
+cv::Size imageSize;
+bool fov_callback_called;
+double angle_constant;
+vector<pcl::PointXYZ> cone_points;
+
+
+void publish_closest_point() {
+    if (cone_points.size() != 0){
+        double min = cone_points[0].x;
+        for (pcl::PointXYZ cone_point : cone_points) {
+            if (cone_point.x < min) {
+                min = cone_point.x;
+            }
+        }
+        std_msgs::Float32 msg;
+        msg.data = (float) min;
+        pub_closest_point.publish(msg);
+    }
+}
+
+
+void fovCallback(const sensor_msgs::CameraInfoConstPtr& msg) {
+    double fx = msg->P[0]; //horizontal focal length of rectified image, in px
+//    fy = msg->P[5]; //vertical focal length
+    imageSize = cv::Size(msg->width, msg->height);
+    camera_fov_horizontal = 2 * atan2(imageSize.width, 2*fx);
+    fov_callback_called = true;
+    angle_constant = camera_fov_horizontal / 2.0 /imageSize.width;
+}
+
+
+void loadCameraFOV(ros::NodeHandle& nh) {
+    auto infoSub = nh.subscribe("/camera/camera_info", 1, fovCallback);
+
+    fov_callback_called = false;
+    while (!fov_callback_called) {
+        ros::spinOnce();
+        ros::Duration(1).sleep();
+    }
+    ROS_INFO("Using horizontal FOV %f ", camera_fov_horizontal);
+}
+
+
+void publishPointCloud(pcl::PointCloud<pcl::PointXYZ> &cloud) {
+    sensor_msgs::PointCloud2 outmsg;
+    pcl::toROSMsg(cloud, outmsg);
+    outmsg.header.frame_id = "base_footprint";
+    pub_pointcloud.publish(outmsg);
+}
+
+
+void drawCircle(pcl::PointCloud<pcl::PointXYZ> &cloud, pcl::PointXYZ point, double radius, int numPoints) {
+    const double pi = boost::math::constants::pi<double>();
+    double angleStep = (2.0 * pi) / numPoints;
+
+    double angle = 0; //start angle
+
+    for (int i = 0; i < numPoints; i++) {
+        //draw a circle
+        double x = radius * cos(angle);
+        double y = radius * sin(angle);
+
+        cloud.push_back(pcl::PointXYZ(point.x + x, point.y + y , 0)); //point.x + radius is center of circle
+        angle = angle + angleStep;
+    }
+}
+
+
+pcl::PointCloud<pcl::PointXYZ> draw_cone_circles() {
+    pcl::PointCloud<pcl::PointXYZ> cone_cloud;
+    for (pcl::PointXYZ cone_point : cone_points) {
+        drawCircle(cone_cloud, cone_point, real_cone_radius, 20);
+    }
+    return cone_cloud;
+}
+
+
+void publishMessage(ros::Publisher pub, Mat img, std::string img_type) {
```

Comment:
Since you have a `publishPointCloud()` may want to call this `publishImage()`

At some point we.may want to create a standard RR library that has stuff like this so you don't have to copy paste every node.